src/core/etl/src/Flow/ETL/PHP/Type/ArrayTypeDetector.php

```diff
@@ -0,0 +1,81 @@
+<?php
+declare(strict_types=1);
+
+namespace Flow\ETL\PHP\Type;
+
+use Flow\ETL\Exception\InvalidArgumentException;
+use Flow\ETL\PHP\Type\Native\ArrayType;
+use Flow\ETL\PHP\Type\Native\NullType;
+use Flow\ETL\PHP\Type\Native\ScalarType;
+
+final class ArrayTypeDetector
+{
+    private readonly array $uniqueKeysType;
+
+    private readonly array $uniqueValuesType;
+
+    public function __construct(
+        array $keysTypes,
+        array $valuesTypes,
+    ) {
+        if (0 === \count($keysTypes)) {
+            throw new InvalidArgumentException('Key type list cannot be empty');
+        }
+
+        if (0 === \count($valuesTypes)) {
+            throw new InvalidArgumentException('Value type list cannot be empty');
+        }
+
+        $this->uniqueKeysType = \array_unique(\array_map(fn (Type $type) : string => \serialize($type), $keysTypes));
+        $this->uniqueValuesType = \array_unique(
+            \array_map(fn (Type $type) : string => \serialize($type), \array_filter($valuesTypes, function (Type $type) : bool {
+                if ($type instanceof NullType) {
+                    return false;
+                }
+
+                return !($type instanceof ArrayType && $type->empty());
```

Comment:
I don't think it should be filtered out here, I would move that filter logic to isMap/isList directly

Reply:
Another thing, that would make this class even easier to read would be to create a class called `Types` that would be a collection of Types with all handy methods on it, like filtering nulls etc

Reply:
One thing here is that we have this filter before serialize, later we would need to unserialize to get the types.

Reply:
Yeah but without that, we won't be able to properly detect if lists/maps are nullable